_posts/2019-03-20-dask-cuml.md

```diff
@@ -0,0 +1,235 @@
+was 19.7s -- this is one area we can improve upon.
+
+## Hyper-Parameter Optimization Experiments
+
+So moving to the GPU can be costly, but once there, with larger data sizes, we gain significant performance
+optimizations.  Naively, we thought, "well, we have GPU machine learning, we have distributed hyper-parameter optimization...
+we *should* have GPU hyper-parameter optimization!"
+
+Sklearn assumes a specific, but well defined API for its object.  Most estimators/classifiers in sklearn look like the following:
+
+```python
+class DummyEstimator(BaseEstimator):
+    def __init__(self, params=...):
+        ...
+
+    def fit(self, X, y=None):
+        ...
+
+    def predict(self, X):
+        ...
+
+    def score(self, X, y=None):
+        ...
+
+    def get_params(self):
+        ...
+
+    def set_params(self, params...):
+        ...
+```
+
+When we started experimenting with hyper-parameter optimization, we found a few API holes missing, these were
+resolved, mostly handling matching argument structure and various getters/setters.
+
+- [#271](https://github.com/rapidsai/cuml/pull/271)
+- [#318](https://github.com/rapidsai/cuml/pull/318)
+- [#330](https://github.com/rapidsai/cuml/pull/330)
+- [#322](https://github.com/rapidsai/cuml/pull/322)
+
+With holes plugged up we tested again.  Using the same diabetes data set, we are now performing hyper-parameter optimization
+and searching over many alpha parameters for the best *scoring* alpha.
+
+```python
+params = {'alpha': np.logspace(-3, -1, 10)}
+clf = linear_model.Ridge(alpha=alpha, fit_intercept=fit_intercept, normalize=normalize, solver='cholesky')
+cu_clf = cumlRidge(alpha=alpha, fit_intercept=fit_intercept, normalize=normalize, solver="eig")
+
+grid = GridSearchCV(clf, params, scoring='r2')
+grid.fit(X_train, y_train)
+
+cu_grid = GridSearchCV(cu_clf, params, scoring='r2')
+cu_grid.fit(X_train, y_train)
+```
+
+Again, reminding ourselves that the data is small ~28KB, we don't expect to observe cuml performing faster than sklearn. Instead, we want to demonstrate functionality.  Fortunately, GridSearchCV from dask-ml adheres to the same API as sklearn; we can swap out GridsearchCV from sklearn and use dask-ml's which will use all of the GPUs we have available.
+
+```python
+params = {'alpha': np.logspace(-3, -1, 10)}
+clf = linear_model.Ridge(alpha=alpha, fit_intercept=fit_intercept, normalize=normalize, solver='cholesky')
+cu_clf = cumlRidge(alpha=alpha, fit_intercept=fit_intercept, normalize=normalize, solver="eig")
+
+grid = dcv.GridSearchCV(clf, params, scoring='r2')
+grid.fit(X_train, y_train)
+
+cu_grid = dcv.GridSearchCV(cu_clf, params, scoring='r2')
+cu_grid.fit(X_train, y_train)
+```
+
+Timing Measurements:
+
+| | sklearn | cuml |
+| -----------| --- | -----|
+| **GridSearchCV** | 88.4 ms ± 6.11 ms | 6.51 s ± 132 |
+| **dask-ml GridSearchCV** | 873 ms ± 347| 740 ms ± 142 |
```

Comment:
Maybe move "GridSearchCV" up to the upper left box, and have "Scikit-Learn" and "Dask-ML" in the row headers.

Reply:
Need units on the +-?